### PR TITLE
Add auto refresh option

### PR DESCRIPTION
**Pull request by akileng56**
- Add refresh to charts

### CODE REVIEW

File: src/BarChart/components/BarChartContainer.ts [hunk 1]
```diff
@@ -23,6 +23,7 @@ export default class BarChartContainer extends Component<BarChartContainerProps,
     };
     private subscriptionHandle?: number;
     private defaultColors: string[] = [ "#2CA1DD", "#76CA02", "#F99B1D", "#B765D1" ];
+    private intervalHandler?: number;
```
Comment (edwinmp) [inline src/BarChart/components/BarChartContainer.ts]: Inaccurate name... It's an interval ID, not handler...

File: src/BarChart/components/BarChartContainer.ts [hunk 2]
```diff
@@ -56,6 +69,9 @@ export default class BarChartContainer extends Component<BarChartContainerProps,
         if (this.subscriptionHandle) {
             mx.data.unsubscribe(this.subscriptionHandle);
         }
+        if (this.intervalHandler) {
+            clearInterval(this.intervalHandler);
```
Comment (edwinmp) [inline src/BarChart/components/BarChartContainer.ts]: Why? This clears the interval when the props change but does not re-create it. So for example in the listening to grid, once a user selects a different context, the chart stops refreshing.

File: src/BarChart/components/BarChartContainer.ts [hunk 3]
```diff
@@ -56,6 +69,9 @@ export default class BarChartContainer extends Component<BarChartContainerProps,
...
+        }
     }
```
Comment (edwinmp) [inline src/BarChart/components/BarChartContainer.ts]: The setInterval is not cleared when the component unmounts.... ??

File: src/BarChart/BarChart.xml [hunk 4]
```diff
@@ -190,6 +190,11 @@
             <category>Appearance</category>
             <description/>
         </property>
+        <property key="refreshInterval" type="integer" defaultValue="0">
+            <caption>Refresh interval (ms)</caption>
+            <category>Appearance</category>
+            <description>Refresh the chart on milliseconds interval, when set to 0 refresh is disabled</description>
```
Comment (edwinmp) [inline src/BarChart/BarChart.xml]: This description sucks. :)

**Review by edwinmp — CHANGES_REQUESTED** [review]

File: src/BarChart/components/BarChartContainer.ts (hunk 1 above)
Comment (akileng56) [reply src/BarChart/components/BarChartContainer.ts]: Changed for all

File: src/BarChart/components/BarChartContainer.ts (hunk 3 above)
Comment (akileng56) [reply src/BarChart/components/BarChartContainer.ts]: Its called after the component did mount which removes ineterval added

File: src/BarChart/BarChart.xml (hunk 4 above)
Comment (akileng56) [reply src/BarChart/BarChart.xml]: Any ideas on a better description??.
The number of milliseconds it takes for a chart to refresh, when set to 0 refresh is disabled

File: src/BarChart/components/BarChartContainer.ts (hunk 2 above)
Comment (akileng56) [reply src/BarChart/components/BarChartContainer.ts]: Added a function call on component will receive props

File: src/HeatMap/components/HeatMapContainer.ts [hunk 5]
```diff
@@ -39,13 +40,29 @@ export default class HeatMapContainer extends Component<HeatMapContainerProps, H
...
         if (this.subscriptionHandle) {
             window.mx.data.unsubscribe(this.subscriptionHandle);
         }
+        if (this.intervalID) {
+            clearInterval(this.intervalID);
+        }
+    }
+
+    private setRefreshInterval(refreshInterval: number) {
+        if (refreshInterval > 0) {
```
Comment (Andries-Smit) [inline src/HeatMap/components/HeatMapContainer.ts]: This function will be called every time the properties are changing. So multiple interfals will be started, overwriting the last ID. causing in the un-mount not be able to clear the interval. causes the internal to function will the component no longer exists.

``` ts
if (this.intervalID) {
    window.clearInterval(this.intervalID);
}
```

Also move the `newProps.mxObject` check to this function. So if the mxobject not available the interval should be cleared.

File: src/HeatMap/components/HeatMapContainer.ts [hunk 6]
```diff
@@ -39,13 +40,29 @@ export default class HeatMapContainer extends Component<HeatMapContainerProps, H
         this.resetSubscriptions(newProps.mxObject);
         if (!this.state.alertMessage) {
             this.fetchData(newProps.mxObject);
+            if (newProps.mxObject) {
+                this.setRefreshInterval(newProps.refreshInterval);
```
Comment (Andries-Smit) [inline src/HeatMap/components/HeatMapContainer.ts]: move `newProps.mxObject` check in to `setRefreshInterval` see comments below

File: src/LineChart/LineChart.xml [hunk 7]
```diff
@@ -199,6 +199,11 @@
             <category>Appearance</category>
             <description/>
         </property>
+        <property key="refreshInterval" type="integer" defaultValue="0">
+            <caption>Refresh interval (ms)</caption>
+            <category>Appearance</category>
```
Comment (Andries-Smit) [inline src/LineChart/LineChart.xml]: `Refresh` is not a real logical property of Appearance. 
Please move `refreshInterval` from all charts into the category `Data source`

File: src/HeatMap/components/HeatMapContainer.ts [hunk 8]
```diff
@@ -39,13 +40,29 @@ export default class HeatMapContainer extends Component<HeatMapContainerProps, H
...
             window.mx.data.unsubscribe(this.subscriptionHandle);
         }
+        if (this.intervalID) {
+            clearInterval(this.intervalID);
+        }
+    }
+
+    private setRefreshInterval(refreshInterval: number) {
+        if (refreshInterval > 0) {
+            this.intervalID = setInterval(() => {
```
Comment (Andries-Smit) [inline src/HeatMap/components/HeatMapContainer.ts]: Global scoped variables should be named explicit.
`window.setInterval`

File: src/HeatMap/components/HeatMapContainer.ts [hunk 9]
```diff
@@ -39,13 +40,29 @@ export default class HeatMapContainer extends Component<HeatMapContainerProps, H
...
+            }
         }
     }
 
     componentWillUnmount() {
         if (this.subscriptionHandle) {
             window.mx.data.unsubscribe(this.subscriptionHandle);
         }
+        if (this.intervalID) {
+            clearInterval(this.intervalID);
```
Comment (Andries-Smit) [inline src/HeatMap/components/HeatMapContainer.ts]: Global scoped variables should be named explicit.
`window.clearInterval`

**Review by Andries-Smit — CHANGES_REQUESTED** [review]
Please update

File: src/HeatMap/components/HeatMapContainer.ts (hunk 5 above)
Comment (akileng56) [reply src/HeatMap/components/HeatMapContainer.ts]: There is a call for compountWillUnmount in the reset subscriptions... So the clear will be done before another is set

File: src/HeatMap/components/HeatMapContainer.ts (hunk 6 above)
Comment (akileng56) [reply src/HeatMap/components/HeatMapContainer.ts]: Changed it

File: src/LineChart/LineChart.xml (hunk 7 above)
Comment (akileng56) [reply src/LineChart/LineChart.xml]: @Andries-Smit The Data source tab is per series and yet this property is for the whole chart. should it be under the tab for Chart Properties then?

File: src/HeatMap/components/HeatMapContainer.ts (hunk 8 above)
Comment (akileng56) [reply src/HeatMap/components/HeatMapContainer.ts]: Added

File: src/HeatMap/components/HeatMapContainer.ts (hunk 9 above)
Comment (akileng56) [reply src/HeatMap/components/HeatMapContainer.ts]: Added

**Pull request closed without merging**